tracing: port USB backend to the new USB device stack

jfischer-no opened the pull request:
Port USB backend to the new USB device stack.

jfischer-no committed:
tracing: port USB backend to the new USB device stack

Port USB backend to the new USB device stack.

Signed-off-by: Johann Fischer <johann.fischer@nordicsemi.no>

jfischer-no committed:
scripts: trace_capture_usb: disable tracing on exit

Disable tracing on exit and do not discard data immediately after
enabling tracing. The length of the data depends on the timings of the
host and targets and may simply be an arbitrary length that does not fit
any trace format or frame.

Signed-off-by: Johann Fischer <johann.fischer@nordicsemi.no>

tmon-nordic commented on subsys/tracing/tracing_backend_usb.c:
```diff
@@ -49,124 +176,113 @@ USBD_CLASS_DESCR_DEFINE(primary, 0) struct usb_device_desc dev_desc = {
 	while (length > 0) {
-		transfer_state = USB_TRANSFER_ONGOING;
-
-		/*
-		 * make sure every USB transfer no need ZLP at all
-		 * because we are in lowest priority thread content
-		 * there are no deterministic time between real USB
-		 * packet and ZLP
-		 */
-		ret = usb_write(TRACING_IF_IN_EP_ADDR, data,
-				length >= CONFIG_TRACING_USB_MPS ?
-				CONFIG_TRACING_USB_MPS - 1 : length, &bytes);
```
If this is a simple port, then why are you changing this behavior?

jfischer-no replied:
The way in which this backend is intended to work is defined by the tracing implementation and the scripts/tracing/trace_capture_usb.py file. The data argument is pointer to ring buffer data, the ring buffer is quite large (CONFIG_TRACING_BUFFER_SIZE). The length argument is the result of ring_buf_get_claim() and the length value is usually much larger than the bulk endpoint MPS. CONFIG_TRACING_USB_MPS is an integer with a default value of 64. When tracing_backend_usb_output() is called, it loops until all the data has been transferred, one step at a time, using short transactions. These transactions are "synchronized" using 'static volatile int transfer_state' (usb_write() is not blocking). On the script side, there is
https://github.com/zephyrproject-rtos/zephyr/blob/8912f070bb0aec621fc49176c46b833eb5e99ece/scripts/tracing/trace_capture_usb.py#L91-L92

This could result in an exact transfer length, a short packet or a timeout. However, in the current implementation, it will always be a short (≤ 63). As long as tracing is enabled there is a tracing data stream (btw, the script enables it but does not disable it on exit and also try do drops some data at beginning, so if one execute the script second time the data is very likely will be corrupted).
I cannot see any reason to handle it using short transactions; the UDC driver can handle the whole transfer more efficiently in an interrupt context, avoiding unnecessary context switching.

carlescufi committed:
scripts: trace_usb_capture: Clean up on exit

When the user presses Ctrl+c to exit the infinite capturing loop, ensure
that the cleanup funciton is always executed.

Signed-off-by: Carles Cufi <carles.cufi@nordicsemi.no>

jfischer-no committed:
samples: tracing: remove USB backend configuration for testing format

Not all the hooks for k_events, k_lifo, k_fifo, k_msgq, and k_queue are
implemented in tracing_test.h and tracing_test.c and therefore the
testing format cannot be used with the new USB device stack.

Signed-off-by: Johann Fischer <johann.fischer@nordicsemi.no>

tmon-nordic commented on subsys/tracing/tracing_backend_usb.c:
```diff
@@ -13,30 +13,160 @@
+	}
+
+	ep = tracing_func_get_bulk_out(c_data);
+	if (USBD_SUPPORTS_HIGH_SPEED &&
+	    usbd_bus_speed(uds_ctx) == USBD_SPEED_HS) {
+		buf = usbd_ep_buf_alloc(c_data, ep, 512);
+	} else {
+		buf = usbd_ep_buf_alloc(c_data, ep, 64);
+	}
```
Is there a need to alloc only 64 bytes at Full-Speed?

It should be possible to just use `USBD_MAX_BULK_MPS` here. The only difference would be that High-Speed capable device operating at Full-Speed would allocate bigger buffer (and thus be able to transfer more data without overhead) but that I consider to be a positive rather than negative (High-Speed capable device have to be able to allocate big buffer in the very likely case it runs at High-Speed).

tmon-nordic commented on scripts/tracing/trace_capture_usb.py:
```diff
@@ -96,6 +92,7 @@ def main():
     except KeyboardInterrupt:
         pass
     finally:
+        write_endpoint.write('disable')
```
Is disable flushing IN data? Is there any synchronization between disable and enable so there is no stale IN data left after the session?

jfischer-no replied:
I quickly looked again at how the asynchronous mode works (what USB backend uses), and I think the backend could try to clean up and dequeue. But TBH, I am not sure if that would prevent outdated frame in the tracing buffer.
